dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/handlers/HttpAuthenticationMechanismHandlerImpl.java

```diff
@@ -0,0 +1,385 @@
+
+        if (multiHttpAuthenticationMechanisms == null) {
+            // first time fetching the auth mechanisms
+            multiHttpAuthenticationMechanisms = new TreeSet<>(priorityComparator);
+            scanAuthenticationMechanisms(multiHttpAuthenticationMechanisms);
+            authMechanismMap.put(moduleName, multiHttpAuthenticationMechanisms);
+        } else if (multiHttpAuthenticationMechanisms.size() > 1) {
+            // Re-sort since priority can change due to deferred EL expressions
+            Set<MultiHttpAuthenticationMechanism> oldMultiHttpAuthenticationMechanisms = multiHttpAuthenticationMechanisms;
+            multiHttpAuthenticationMechanisms = new TreeSet<>(priorityComparator);
+            multiHttpAuthenticationMechanisms.addAll(oldMultiHttpAuthenticationMechanisms);
+        }
+
+        if (multiHttpAuthenticationMechanisms.isEmpty()) {
+            if (tc.isDebugEnabled()) {
+                Tr.debug(tc, "No HttpAuthenticationMechanism found");
+            }
+
+            Tr.error(tc, "JAKARTASEC_ERROR_NO_HAM", getModuleName(), getApplicationName());
+            return null;
+        }
+
+        logHttpAuthenticationMechanisms(multiHttpAuthenticationMechanisms);
+        checkUniquePriorityFound(multiHttpAuthenticationMechanisms);
+
+        // Return the highest priority mechanism (first in the sorted set)
+        return multiHttpAuthenticationMechanisms.iterator().next();
+    }
+
+    /**
+     * Ensure a unique HAM is found and throw an exception if one cannot be found.
+     *
+     * All in-built HAMs will have a unique, ordered priority, so this only applies
+     * to if there are more than one application HAMs, and the top 2 have the same
+     * priority (i.e. can't pick between them).
+     *
+     * @param multiHttpAuthenticationMechanisms
+     */
+    private void checkUniquePriorityFound(Set<MultiHttpAuthenticationMechanism> multiHttpAuthenticationMechanisms) {
+        // if we have fewer than 2 mechanisms, there's no ambiguity
+        if (multiHttpAuthenticationMechanisms.size() < 2) {
+            return;
+        }
+
+        // get the first two mechanisms from the set ...
+        Iterator<MultiHttpAuthenticationMechanism> iterator = multiHttpAuthenticationMechanisms.iterator();
+        MultiHttpAuthenticationMechanism first = iterator.next();
+        MultiHttpAuthenticationMechanism second = iterator.next();
+
+        // ... and compare their priorities
+        if (first.getPriority() == second.getPriority()) {
+            if (tc.isDebugEnabled()) {
+                Tr.debug(tc, "Found two HttpAuthenticationMechanisms with the same priority: "
+                             + first.getSimpleName() + " and " + second.getSimpleName()
+                             + " both with priority " + first.getPriority());
+            }
+
+            Tr.error(tc, "JAKARTASEC_ERROR_AMBIGUOUS_RESOLUTION", first.getSimpleName(), second.getSimpleName(), first.getPriority());
```

Comment:
In relation to my comments previously in the messages file for this error. Is this the only place we output this error? if it is, then we should focus the message on priority and not mention other potential factors, given we don't seem to check for anything within our handler. 

I assume other factors would be for customers own provided Multiple HAM handler where they will provide their own errors/exceptions/messages. Customers do not reuse our messages within their own code or user features